ce/api/streamflow/watershed.py

```diff
@@ -0,0 +1,421 @@
+        ]
+
+
+def watershed(sesh, station, ensemble_name):
+    """Return information describing the watershed that drains to a specified
+    point.
+
+    :param sesh: (sqlalchemy.orm.session.Session) A database Session object
+    :param station: (string) Location of drainage point, WKT POINT format
+    :param ensemble_name: (string) Name of the ensemble containing data files backing
+        providing data for this request.
+    :return: (dict) representation for JSON response object. See function
+        `worker` for details.
+
+    This function is primarily responsible for finding the relevant data files
+    and converting their contents to `DataGrid` objects for consumption by
+    `worker`, which as its name suggests, does most of the work.
+    """
+    station_lonlat = WKT_point_to_lonlat(station)
+
+    with get_time_invariant_variable_dataset(
+        sesh, ensemble_name, 'flow_direction'
+    ) as flow_direction_ds:
+        with get_time_invariant_variable_dataset(
+            sesh, ensemble_name, 'elev'
+        ) as elevation_ds:
+            with get_time_invariant_variable_dataset(
+                sesh, ensemble_name, 'area'
+            ) as area_ds:
+                try:
+                    return worker(
+                        station_lonlat,
+                        flow_direction=DataGrid.from_nc_dataset(flow_direction_ds, 'flow_direction'),
+                        elevation=DataGrid.from_nc_dataset(elevation_ds, 'elev'),
+                        area=DataGrid.from_nc_dataset(area_ds, 'area'),
+                    )
+                except DataGridIndexError:
+                    abort(404, description=
+                        'Station lon-lat coordinates are not within the area '
+                        'for which we have data')
+
+
+def worker(station_lonlat, flow_direction, elevation, area, hypso_params=None):
+    """Compute the watershed endpoint response.
+
+    This function exists to make these computations more easily testable.
+    (Specifically, data *files* are not required, only the relevant contents
+    of those files passed as `DataGrid` objects. `DataGrid`s are much easier to
+    construct for tests.)
+
+    :param station_lonlat: (tuple) Location of drainage point, (lon, lat)
+    :param flow_direction: (DataGrid) Flow direction grid
+    :param elevation: (DataGrid) Elevation grid
+    :param area: (DataGrid) Area grid
+    :return: dict representation for JSON response object with the following
+        attributes:
+            area: Area of the watershed
+            elevation: Minimum and maximum elevations
+            shape: A GeoJSON object representing the outline of the watershed;
+                a concave hull of the cell rectangles.
+            hypsometric_curve: Elevation-area histogram of the watershed
+    """
+    if hypso_params is None:
+        # Default parameters cover total range of BC elevations from
+        # sea level to highest point, Mt. Fairweather, at 4,663 m.
```

Comment:
Fine for now, but [Mount Logan](https://en.wikipedia.org/wiki/Mount_Logan) isn't far afield of BC's borders. Flag this if we start modelling the coast range. Rainier and Hood are within this range, so we're probably good for the Columbia.